### PR TITLE
Add new transformation to clang_delta

### PR DESCRIPTION
**Pull request by mortior**
I had some problems with "fully" reduced files with a lot of unnecessary array accesses in them so I added a transformation that replaces array-accesses with their index expression.

### CODE REVIEW

File: clang_delta/Makefile.am [hunk 1]
```diff
@@ -197,6 +197,8 @@ clang_delta_SOURCES = \
 	RenameVar.h \
 	ReplaceArrayIndexVar.cpp \
 	ReplaceArrayIndexVar.h \
+	ReplaceArrayAccessWithIndex.cpp \
+	ReplaceArrayAccessWithIndex.h \
```
Comment (chenyang78) [inline clang_delta/Makefile.am]: Move these two lines above ReplaceArrayIndexVar.cpp to maintain lexical order?

File: clang_delta/ReplaceArrayAccessWithIndex.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+struct ReplaceArrayAccessWithIndex::IndexCollector
+  : public RecursiveASTVisitor<ReplaceArrayAccessWithIndex::IndexCollector>
+{
+  IndexCollector(ReplaceArrayAccessWithIndex *instance);
+  bool VisitArraySubscriptExpr(ArraySubscriptExpr *ASE);
+
+  const VarDecl *getVarDeclFromExpr(const Expr *E);
+
+  ReplaceArrayAccessWithIndex *ConsumerInstance;
```
Comment (chenyang78) [inline clang_delta/ReplaceArrayAccessWithIndex.cpp]: Sorry for being pedantic - can we make it ```private``` by following the existing convention? Thanks.

File: clang_delta/ReplaceArrayAccessWithIndex.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+static const char *Description =
+  "Replace array accesses with the index expression.";
+
+static RegisterTransformation<ReplaceArrayAccessWithIndex>
+Trans("replace-array-access-with-index", Description);
+
+struct ReplaceArrayAccessWithIndex::IndexCollector
+  : public RecursiveASTVisitor<ReplaceArrayAccessWithIndex::IndexCollector>
+{
+  IndexCollector(ReplaceArrayAccessWithIndex *instance);
```
Comment (chenyang78) [inline clang_delta/ReplaceArrayAccessWithIndex.cpp]: Make the constructor ```explicit```?

File: clang_delta/ReplaceArrayAccessWithIndex.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+bool ReplaceArrayAccessWithIndex::IndexCollector::VisitArraySubscriptExpr(
+  ArraySubscriptExpr *ASE)
+{
+  // Skip expressions in included files.
+  if (ConsumerInstance->isInIncludedFile(ASE))
+    return true;
+
+  // Skip multi-dimensional arrays.
```
Comment (chenyang78) [inline clang_delta/ReplaceArrayAccessWithIndex.cpp]: Perhaps the comment needs to be attached to line 59 below (i.e., ``` if (!ArrayTy || ...)``` )?

File: clang_delta/ReplaceArrayAccessWithIndex.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  const VarDecl *BaseVD = getVarDeclFromExpr(ASE->getBase());
+
+  if (!BaseVD)
+    return true;
+
+  ArrayType const *ArrayTy = dyn_cast<ArrayType>(BaseVD->getType().getTypePtr());
+
+  if (!ArrayTy || dyn_cast<ArrayType>(ArrayTy->getElementType().getTypePtr()))
```
Comment (chenyang78) [inline clang_delta/ReplaceArrayAccessWithIndex.cpp]: Can we also skip cases where the array element type is incompatible with integer type? For example, with the current implementation, this transformation turns the code below

    struct S { int f; };
    int foo() {
      S s[1] = {{1}};
      return s[0].f;
    }

into the following form that can't be compiled:

    struct S { int f; };
       int foo() {
      S s[1] = {{1}};
      return (struct S)(0).f;
    }

Adding more restrictions can reduce the chances of generating "bad" output.

File: clang_delta/ReplaceArrayAccessWithIndex.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
+#ifndef REPLACE_ARRAY_ACCESS_WITH_INDEX
+#define REPLACE_ARRAY_ACCESS_WITH_INDEX
+
+#include "Transformation.h"
+
+class ReplaceArrayAccessWithIndex : public Transformation {
+public:
+
+  ReplaceArrayAccessWithIndex(const char *TransName, const char *Desc)
+    : Transformation(TransName, Desc)
```
Comment (chenyang78) [inline clang_delta/ReplaceArrayAccessWithIndex.h]: Initialize ```Collector``` to ```NULL```?

**Comment by chenyang78** [comment]
If possible, could you include the new Makefile.in re-generated by automake 1.15, which is the version that we used for generating the current Makefile.in. Thank for the patch, Johan!

**Comment by mortior** [comment]
I think all comments have been taken care of.

**Comment by chenyang78** [comment]
Hi Johan,

Sorry for my late response! I think the pass is good at this point, so I will merge it. If you have time, it would be great if you could make the pass avoid the bad transformations on the following code:

    $ cat array.cc
    void foo() {
      int a[10];
      int *p = &a[0];
      a[0] = 1;
    }
 
    $ clang_delta --transformation=replace-array-access-with-index --counter=1 array.cc 
    void foo() {
      int a[10];
      int *p = &0;
      a[0] = 1;
    }

    $ clang_delta --transformation=replace-array-access-with-index --counter=2 array.cc
    void foo() {
      int a[10];
      int *p = &a[0];
      0 = 1;
    }

You can see that the transformed programs won't be compiled successfully. 

Again, thanks for contributing the patch!

**Comment by eeide** [comment]
@mortier, please read issue #99, which I just opened regarding the (missing) copyright and license notices in the files that implement the new pass.

**Pull request merged**